src/core/utils/channel_monitor.cpp

```diff
@@ -181,13 +189,16 @@ void ChannelMonitor::HandleEnergyScanResult(otEnergyScanResult *aResult)
 
 void ChannelMonitor::LogResults(void)
 {
-    otLogInfoUtil(
-        "ChannelMonitor: %u [%02x %02x %02x %02x %02x %02x %02x %02x %02x %02x %02x %02x %02x %02x %02x %02x]",
-        mSampleCount, mChannelOccupancy[0] >> 8, mChannelOccupancy[1] >> 8, mChannelOccupancy[2] >> 8,
-        mChannelOccupancy[3] >> 8, mChannelOccupancy[4] >> 8, mChannelOccupancy[5] >> 8, mChannelOccupancy[6] >> 8,
-        mChannelOccupancy[7] >> 8, mChannelOccupancy[8] >> 8, mChannelOccupancy[9] >> 8, mChannelOccupancy[10] >> 8,
-        mChannelOccupancy[11] >> 8, mChannelOccupancy[12] >> 8, mChannelOccupancy[13] >> 8, mChannelOccupancy[14] >> 8,
-        mChannelOccupancy[15] >> 8);
+    const size_t kStringSize = 128;
+    char         buf[kStringSize];
+    char *       cur = buf;
+
+    for (size_t i = 0; i < kNumChannels; i++)
+    {
+        cur += snprintf(cur, sizeof(buf) - static_cast<size_t>(cur - buf), "%02x ", mChannelOccupancy[i] >> 8);
```

Comment:
This works ok with `buf` being 128 char and us having max of 26 channels.

But in general, I suggest not using `cur += snprintf(..)` model. `snprintf()` returns  the number of characters that would have been written if buffer had been sufficiently large,  and not the number of chars that are actually written ([link](http://www.cplusplus.com/reference/cstdio/snprintf/)). So it's possible to go beyond the end of a buffer.

How about using `ot::String<kStringSize>` instead? We can use [`String::Append()`](https://github.com/openthread/openthread/blob/master/src/core/common/string.hpp#L181) which appends `printf` style formatted content to the string and ensures we never go beyond end of buffer.

Reply:
At my first implementation, this code is using `ot::String<kStringSize>`  to format the string. But @bukepo gives me a comment said `ot::String<kStringSize>` will increase lots of code space. So I did an experiment to verify his comment:
<1>Function LogResults is the same with the original version:
   text	   data	    bss	    dec	    hex	filename
 180784	    548	  46928	 228260	  37ba4	ot-cli-ftd

<2>Function LogResults call ot::String<kStringSize>:
   text	   data	    bss	    dec	    hex	filename
 180936	    548	  46928	 228412	  37c3c	./output/nrf52840/bin/ot-cli-ftd
Code size increases 152 bytes compared with <1>.

<3>Function LogResults call snprintf:
   text	   data	    bss	    dec	    hex	filename
 180840	    548	  46928	 228316	  37bdc	./output/nrf52840/bin/ot-cli-ftd
Code size increases 56 bytes compared with <1>.
 
So if the function is using `snprintf`, it will save about 100 bytes. How should I choose?

Reply:
Ah. Did not know that. :).

Thanks for all the info and analysis.

We can consider using a string size which is already in use by another module. For example in `mac_frame.hpp`, the class `Mac::Frame` uses a string with size `110` already. If we use `String<110>` here as well then the template instance code should be shared and the code size should hopefully increase less.

At the end, in my opinion, code-safety is worth 100 bytes code increase (particularly when we consider the total image is ~180K, so 100 bytes is about 0.06%).

Thoughts?

Reply:
I did a quick check of this with cc2538 build. Using same length `String<110>` as `mac_frame` should help. 
```
Enabled channel-monitor

(Branch as is)
arm-none-eabi-size  output/cc2538/bin/ot-ncp-ftd
   text	   data	    bss	    dec	    hex	filename
 237564	   5616	  22844	 266024	  40f28	output/cc2538/bin/ot-ncp-ftd

(With the String<110>)
 arm-none-eabi-size  output/cc2538/bin/ot-ncp-ftd
   text	   data	    bss	    dec	    hex	filename
 237560	   5616	  22844	 266020	  40f24	output/cc2538/bin/ot-ncp-ftd
 (code is smaller -4 bytes)

(With String<128>)
 arm-none-eabi-size  output/cc2538/bin/ot-ncp-ftd
   text	   data	    bss	    dec	    hex	filename
 237596	   5616	  22844	 266056	  40f48	output/cc2538/bin/ot-ncp-ftd
(code +32 bytes)
```

Reply:
Agree with you. If the code does not occupy a lot of space, it's better to use safer code.